debugger/loadprinter.ml

```diff
@@ -91,46 +91,47 @@ let eval_value_path env path =
 
 (* Install, remove a printer (as in toplevel/topdirs) *)
 
-let match_printer_type desc make_printer_type =
-  Ctype.with_local_level_generalize begin fun () ->
-    let ty_arg = Ctype.newvar() in
-    Ctype.unify Env.empty
-      (make_printer_type ty_arg)
-      (Ctype.instance desc.val_type);
-    ty_arg
-  end
-
-let find_printer_type lid =
+(* Nearly identical to Topdirs.find_printer
+   (raises exceptions instead of returning a result) *)
+let find_printer lid =
   match Env.find_value_by_name lid Env.empty with
-  | (path, desc) -> begin
-      match match_printer_type desc Topprinters.printer_type_new with
-      | ty_arg -> (ty_arg, path, false)
-      | exception Ctype.Unify _ -> begin
-          match match_printer_type desc Topprinters.printer_type_old with
-          | ty_arg -> (ty_arg, path, true)
-          | exception Ctype.Unify _ -> raise(Error(Wrong_type lid))
-        end
-    end
   | exception Not_found ->
       raise(Error(Unbound_identifier lid))
-
+  | (path, desc) ->
+      match Topprinters.match_printer_type Env.empty desc.val_type with
+      | None -> raise(Error(Wrong_type lid))
+      | Some kind -> (path, kind)
```

Comment:
The `Toplevel.find_printer` function could be moved to `Topprinter` too and modified to return a sum type in the error case. This would make it straightforward to reuse it here.

Reply:
Not sure I'm a huge fan of the refactoring of `Loadprinter.Error` it requires but done.

Reply:
I agree that I don't like the refactoring of `Loadprinter.Error`. In this direction, I would rather suggest defining a more precise datatype for `find_printer` error (for example a polymorphic variant, or just a dedicated type of exceptions), and translate this error information in both the debugger and the toplevel into their preferred (different) error-description format.

Pragmatically, I would be happy to approve and merge the first three commits of this PR, so you could also give up on this suggestion by Florian for the present PR, and rework that part independently.

Reply:
I've implemented the polymorphic variant option. This is clearly better (than my previous attempt, no opinion with respect to "a dedicated type of exceptions").